hudi-common/src/main/java/org/apache/hudi/common/util/ParquetUtils.java

```diff
@@ -105,6 +121,111 @@
     return rowKeys;
   }
 
+  /**
+   * Fetch records from the given parquet file as {@link List} of {@link ParquetRowInfo}.
+   *
+   * @param filePath      The parquet file path.
+   * @param configuration configuration to build fs object
+   * @return {@link List} of {@link ParquetRowInfo}
+   */
+  public static List<ParquetRowInfo> fetchRecordKeyPartitionPathFromParquet(Configuration configuration, Path filePath,
+                                                                            String baseInstantTime,
+                                                                            String fileId) {
+    return fetchRecordKeyPartitionPathFromParquet(configuration, filePath, baseInstantTime, fileId, Collections.EMPTY_LIST);
+  }
+
+  /**
+   * Fetch records from the given parquet file as {@link List} of {@link ParquetRowInfo}. Filter for candidates if set.
+   *
+   * @param filePath      The parquet file path.
+   * @param configuration configuration to build fs object
+   * @return {@link List} of {@link ParquetRowInfo} matching candidateRecordKeys
+   */
+  static List<ParquetRowInfo> fetchRecordKeyPartitionPathFromParquet(Configuration configuration, Path filePath,
+                                                                     String baseInstantTime,
+                                                                     String fileId, List<String> candidatesToFilter) {
+    List<ParquetRowInfo> parquetRowInfos = new ArrayList<>();
+    try {
+      if (!filePath.getFileSystem(configuration).exists(filePath)) {
+        return new ArrayList<>();
+      }
+
+      Configuration conf = new Configuration(configuration);
+      conf.addResource(FSUtils.getFs(filePath.toString(), conf).getConf());
+      Schema readSchema = HoodieAvroUtils.getRecordKeyPartitionPathSchema();
+      AvroReadSupport.setAvroReadSchema(conf, readSchema);
+      AvroReadSupport.setRequestedProjection(conf, readSchema);
+      ParquetReader reader = AvroParquetReader.builder(filePath).withConf(conf).build();
+      Object obj = reader.read();
+      while (obj != null) {
+        if (obj instanceof GenericRecord) {
+          String recordKey = ((GenericRecord) obj).get(HoodieRecord.RECORD_KEY_METADATA_FIELD).toString();
+          String partitionPath = ((GenericRecord) obj).get(HoodieRecord.PARTITION_PATH_METADATA_FIELD).toString();
+          if (candidatesToFilter.isEmpty() || candidatesToFilter.contains(recordKey)) {
+            parquetRowInfos.add(new ParquetRowInfo(recordKey, partitionPath, baseInstantTime, fileId));
+          }
+          obj = reader.read();
+        }
+      }
+    } catch (IOException e) {
+      throw new HoodieIOException("Failed to read from Parquet file " + filePath, e);
+    }
+    return parquetRowInfos;
+  }
+
+  /**
+   * Class to hold info for a row in Parquet like record key, partition path, base instance time and fileId.
+   */
+  public static class ParquetRowInfo {
```

Comment:
I think we should remove this class and just reuse `HoodieKey`, which is what we actually read out of the parquet file.. the `baseInstantTime` and `fileId` are actually passed in by the caller.. there is no need to actually do this at this level.. We can just reuse the values we pass in at the read handle calling code